### PR TITLE
ENH: add Decimal support to numpy.lib.financial

### PR DESCRIPTION
**Pull request by garrypolley**
Adds support to rate, pv, fv, pmt, ppmt, ipmt, mirr, npv functions when using Decimal type.

Closes #9781

I added a helper method and updated a few others to keep the code consistent with how it had existed before I changed it.

I'm not a huge fan of the `isinstance(VAR, Decimal)` but I can't think of a safer way to do the casting of objects to `Decimal`. This casting keeps the numbers from becoming floats by default; which causes lots of issues with `Decimal` type.

### CODE REVIEW

File: numpy/lib/financial.py [hunk 1]
```diff
@@ -22,17 +24,20 @@
                 'start':1,
                 'finish':0}
 
-def _convert_when(when):
+def _convert_when(when, sentinel=None):
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: Why are you changing the `when` argument here?

File: numpy/lib/financial.py [hunk 2]
```diff
@@ -271,20 +279,20 @@ def nper(rate, pmt, pv, fv=0, when='end'):
...
     with np.errstate(divide="raise"):
         try:
-            z = pmt*(1.0+rate*when)/rate
+            z = pmt*(1+rate*when)/rate
         except FloatingPointError:
             use_zero_rate = True
 
     if use_zero_rate:
-        return (-fv + pv) / (pmt + 0.0)
+        return (-fv + pv) / (pmt + 0)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: +0 seems pointless now

File: numpy/lib/financial.py (hunk 1 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: I'm changing it to ensure that the called/cached values are always of the correct type. 

Since this function is called in almost all of the financial functions this was the easiest way to ensure all of them used the correct where values. 

The real issue is that if the `where_num`, as I've called it, is not a Decimal type, then the calculations are broken later because it won't be `Decimal('0')` or `Decimal('1')`. It'll end up being `1.0`, or `0.0` in the calculations, which then causes a type error because `Decimal` and `float` are not friends.

File: numpy/lib/financial.py (hunk 2 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: I concur, I'll remove all occurences in the file.

File: numpy/lib/financial.py [hunk 3]
```diff
@@ -206,12 +211,12 @@ def pmt(rate, nper, pv, fv=0, when='end'):
...
-    when = _convert_when(when)
+    when = _convert_when(when, sentinel=rate)
     (rate, nper, pv, fv, when) = map(np.array, [rate, nper, pv, fv, when])
     temp = (1 + rate)**nper
-    mask = (rate == 0.0)
-    masked_rate = np.where(mask, 1.0, rate)
-    z = np.zeros(np.broadcast(masked_rate, nper, pv, fv, when).shape)
+    mask = (rate == 0)
+    masked_rate = np.where(mask, 1, rate)
+    z = np.zeros(np.broadcast(masked_rate, nper, pv, fv, when).shape, dtype=rate.dtype)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: Should probably be `np.result_type(rate, pv, fv)`

File: numpy/lib/financial.py [hunk 4]
```diff
@@ -727,12 +744,17 @@ def mirr(values, finance_rate, reinvest_rate):
         Modified internal rate of return
 
     """
-    values = np.asarray(values, dtype=np.double)
+    values = np.asarray(values)
     n = values.size
+
+    # Support decimal computation
+    if isinstance(finance_rate, Decimal):
+        n = Decimal(n)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: Why is this needed?

Comment (garrypolley) [reply numpy/lib/financial.py]: If `n` is not a decimal then the computation below `1/(n - 1)` becomes a float and causes a `TypeError`. 

`TypeError: unsupported operand type(s) for ** or pow(): 'decimal.Decimal' and 'float'`

File: numpy/lib/financial.py (hunk 3 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: Should it match the args given to the `np.broadcast` call? 

So it'd be: 

```python
z = np.zeros(np.broadcast(masked_rate, nper, pv, fv, when).shape,
             dtype=np.result_type(masked_rate, nper, pv, fv, when))
```

File: numpy/lib/financial.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/lib/financial.py]: Hmm, you're right. I can't think of a good way around that.

On the other hand, you already say some other functions are not supported because they use `sqrt` - if those aren't, then this shouldn't be either

File: numpy/lib/tests/test_financial.py [hunk 5]
```diff
@@ -147,6 +215,79 @@ def test_when(self):
...
+        assert_almost_equal(np.ipmt(Decimal('0.1')/Decimal('12'), Decimal('1'), Decimal('24'), Decimal('2000'),
+                                    Decimal('0'), Decimal('1')).flat[0],
+                            np.ipmt(Decimal('0.1') / Decimal('12'), Decimal('1'), Decimal('24'), Decimal('2000'),
+                                    Decimal('0'), 'begin').flat[0], 2)
+        # end
+        assert_almost_equal(np.ipmt(Decimal('0.1')/Decimal('12'), Decimal('1'), Decimal('24'), Decimal('2000'),
+                                    Decimal('0')).flat[0],
+                            np.ipmt(Decimal('0.1')/Decimal('12'), Decimal('1'), Decimal('24'), Decimal('2000'),
+                                    Decimal('0'), 'end').flat[0], 2)
+        assert_almost_equal(np.ipmt(Decimal('0.1')/Decimal('12'), Decimal('1'), Decimal('24'), Decimal('2000'),
```
Comment (eric-wieser) [inline numpy/lib/tests/test_financial.py]: Does  `assert_almost_equal` cast the decimal to float first? And what's the point of using `Decimal` if you're only going to check for approximate equality?

Comment (garrypolley) [reply numpy/lib/tests/test_financial.py]: I suppose I could change them much like I did in one of the test to check the specific values. 

As for the approximate equality, that is because of the `assert_almost_equal`

Is it alright if I make the class inherit from `unittest.TestCase` and use the `self.assertEquals`? that's really what I want anyway. I was avoiding adding too much new. 

I could also use `assert` directly, which will raise the assertion error, but won't have nice messaging when the test fails by default.

File: numpy/lib/financial.py (hunk 4 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: Is there value in not supporting `Decimal` in places that it can be supported? That's why I've added to everywhere that I could. 

I couldn't figure out how/where `.log` and the eigen log value calculations were located. I would have added the support in those places to call the appropriate Decimal methods, however, I couldn't find it, so I couldn't add it.

**Comment by garrypolley** [comment]
I've updated based on the code review comments. Hopefully the use of `unittest.TestCase` does not break older versions of Python.  The CI shall let us know.

**Comment by eric-wieser** [comment]
Why do you need `unittest.TestCase`? Can't you use `np.testing.assert_equal`?

File: numpy/lib/financial.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/lib/financial.py]: But id doesn't end up as `0.0` or `1.0`, does it? It returns an integer instead.

File: numpy/lib/financial.py [hunk 6]
```diff
@@ -206,12 +211,13 @@ def pmt(rate, nper, pv, fv=0, when='end'):
...
+    when = _convert_when(when, sentinel=rate)
     (rate, nper, pv, fv, when) = map(np.array, [rate, nper, pv, fv, when])
     temp = (1 + rate)**nper
-    mask = (rate == 0.0)
-    masked_rate = np.where(mask, 1.0, rate)
-    z = np.zeros(np.broadcast(masked_rate, nper, pv, fv, when).shape)
+    mask = (rate == 0)
+    masked_rate = np.where(mask, 1, rate)
+    z = np.zeros(np.broadcast(masked_rate, nper, pv, fv, when).shape,
+                 dtype=np.result_type(masked_rate, nper, pv, fv, when))
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: Unrelated to this patch, but: I'd question whether the broadcast or allocation is even necessary here - `where ` does its own broadcasting.

I suspect that `z = 0` will work fine here!

File: numpy/lib/financial.py [hunk 7]
```diff
@@ -220,6 +226,9 @@ def nper(rate, pmt, pv, fv=0, when='end'):
     """
     Compute the number of periodic payments.
 
+    Decimal type is not supported for values due to the use of log computations within
+    the calculation(s).
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: This should be possible to fix - right now, it calls `Decimal.log`, but that doesn't exist. The fix would be to modify something in `generate_umath` which currently attaches `PyUFunc_O_O_method` for `dtype=object`, to allow multiple method names to be tried. Can you file an issue for fixing this?

File: numpy/lib/financial.py [hunk 8]
```diff
@@ -271,20 +280,20 @@ def nper(rate, pmt, pv, fv=0, when='end'):
...
     if use_zero_rate:
-        return (-fv + pv) / (pmt + 0.0)
+        return (-fv + pv) / pmt
     else:
-        A = -(fv + pv)/(pmt+0.0)
-        B = np.log((-fv+z) / (pv+z))/np.log(1.0+rate)
+        A = -(fv + pv)/(pmt+0)
+        B = np.log((-fv+z) / (pv+z))/np.log(1+rate)
         miter = np.broadcast(rate, pmt, pv, fv, when)
         zer = np.zeros(miter.shape)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: I bet `zer = 0` works here too

File: numpy/lib/financial.py [hunk 9]
```diff
@@ -502,11 +511,11 @@ def pv(rate, nper, pmt, fv=0.0, when='end'):
     need initial investments of $649.27 and $1273.79, respectively.
 
     """
-    when = _convert_when(when)
+    when = _convert_when(when, sentinel=rate)
     (rate, nper, pmt, fv, when) = map(np.asarray, [rate, nper, pmt, fv, when])
     temp = (1+rate)**nper
     miter = np.broadcast(rate, nper, pmt, fv, when)
-    zer = np.zeros(miter.shape)
+    zer = np.zeros(miter.shape, dtype=rate.dtype)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: `zer = 0`

File: numpy/lib/financial.py [hunk 10]
```diff
@@ -572,8 +581,14 @@ def rate(nper, pmt, pv, fv, when='end', guess=0.10, tol=1e-6, maxiter=100):
     OpenDocument-formula-20090508.odt
 
     """
-    when = _convert_when(when)
+    when = _convert_when(when, sentinel=pmt)
+    # Cast the default floats to decimal to support decimal computations
+    if isinstance(pmt, Decimal):
+        guess = Decimal(guess)
+        tol = Decimal(tol)
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: This also promote `float`s which _aren't_ the default, which is incorrect. This should error, IMO:

    rate(nper, pmt_decimal, pv, fv, guess=0.20)

for the same reason that python makes `Decimal(1) + 2.0` error

So maybe change the defaults to `np.NoValue_`, and then set the appropriate floats/Decimals in the function? (and add the defaults to the docstring)

File: numpy/lib/financial.py [hunk 11]
```diff
@@ -597,6 +612,9 @@ def irr(values):
     that gives a net present value of 0.0; for a more complete explanation,
     see Notes below.
 
+    Decimal type is not supported for values due to the use of square roots within
+    the calculation(s).
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: Really? `np.sqrt(Decimal(2))` works for me.

File: numpy/lib/financial.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/lib/financial.py]: What do you mean by "eigen log value"?

I thought the problem was that python doesn't even define `log` on `Decimal` deliberately, but that's not the case. Just keep in mind the operations that python does and doesn't allow, and try to respect them. What you have here is fine

Could you expand this python comment explaining it's for the (1/n-1) below?

File: numpy/lib/tests/test_financial.py [hunk 12]
```diff
@@ -7,35 +9,48 @@
...
+        self.assertAlmostEqual(np.irr(v), 0.0886, 2)
 
         # Test that if there is no solution then np.irr returns nan
         # Fixes gh-6744
         v = [-1, -2, -3]
         assert_equal(np.irr(v), np.nan)
 
     def test_pv(self):
-        assert_almost_equal(np.pv(0.07, 20, 12000, 0), -127128.17, 2)
+        self.assertAlmostEqual(np.pv(0.07, 20, 12000, 0), -127128.17, 2)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_financial.py]: Please change these back - we recently stopped using `unittest.TestCase`, due to a planned move to pytest. You can use `assert_equal` in place of `assert_almost_equal`

File: numpy/lib/tests/test_financial.py [hunk 13]
```diff
@@ -51,101 +66,241 @@ def test_pmt(self):
...
+        self.assertEqual(res, tgt)
+        # Test the case where we use broadcast and
+        # the arguments passed in are arrays.
+        res = np.pmt([[Decimal('0'), Decimal('0.8')], [Decimal('0.3'), Decimal('0.8')]],
+                     [Decimal('12'), Decimal('3')], [Decimal('2000'), Decimal('20000')])
+        tgt = np.array([[Decimal('-166.6666666666666666666666667'), Decimal('-19311.25827814569536423841060')],
+                        [Decimal('-626.9081401700757748402586600'), Decimal('-19311.25827814569536423841060')]])
+
+        # Cannot use the `assert_allclose` because it uses isfinite under the covers
+        # which does not support the Decimal type
```
Comment (eric-wieser) [inline numpy/lib/tests/test_financial.py]: This could do with an issue being reported (and then mentioned here) :)

**Review by eric-wieser — COMMENTED** [review]
This is looking pretty close to done. Can you remove the `_convert_when` changes, and see what breaks? I'd rather the fix was local for each one.

File: numpy/lib/financial.py (hunk 4 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: I plan to make all the changes as mentioned, in this review. I appreciate the fast feedback.

I think the call to log and square root will work out fine once the ticket for using `round`, `log`, etc. for different types works. 

It seems decimal does have `.log*` methods that'd mostly do the job, but it'd have to be custom for Decimal because it has limited log support.

File: numpy/lib/tests/test_financial.py (hunk 12 above)
Comment (garrypolley) [reply numpy/lib/tests/test_financial.py]: 😄  I did not know the `assert_equal` was in there as a helper and I'm used to using the `unitest.TestCase`. I'll make the updates, thanks for pointing it out.

File: numpy/lib/financial.py (hunk 11 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: I misrepresented this statement. It's because of the call to np.roots and it returning imaginary numbers. 

Currently I do not think the underlying `eigvals` call within the np.roots method will work because the polynomial call in `np.roots` does a force casts to floats. 

I've given it a try to not force cast to float, and keep the `object` type for dtype. This causes later problems inside of the `eigvals` call (called form `np.roots`). These lines do not support `Decimal` and I'm not sure how to add it: https://github.com/numpy/numpy/blob/v1.13.3/numpy/linalg/linalg.py#L894-L895, there is no matching signature that can be called.

File: numpy/lib/financial.py (hunk 7 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: > Can you file an issue for fixing this?

I will do that today.

File: numpy/lib/financial.py (hunk 11 above)
Comment (eric-wieser) [reply numpy/lib/financial.py]: Yeah, sounds like you're on a loser there. Maybe just fix the comment to refer to `` `roots` ``.

File: numpy/lib/financial.py (hunk 7 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: Done: https://github.com/numpy/numpy/issues/9955

File: numpy/lib/financial.py (hunk 11 above)
Comment (garrypolley) [reply numpy/lib/financial.py]: I've not logged a ticket for this issue because I think it requires a lot more work than a single ticket to end up being supported, and relates heavily to the effort to get `Decimal` to have greater support within numpy overall.

**Comment by garrypolley** [comment]
I've updated from all the PR comments. Let me know what else, if anything, needs updates or further issues logged.

**Comment by garrypolley** [comment]
I've rebased for better/easier merging.

File: numpy/lib/financial.py [hunk 14]
```diff
@@ -573,15 +570,26 @@ def rate(nper, pmt, pv, fv, when='end', guess=0.10, tol=1e-6, maxiter=100):
...
+    if guess is None:
+        guess = default_type('0.1')
+
+    if tol is None:
+        tol = default_type('1e-6')
+
     (nper, pmt, pv, fv, when) = map(np.asarray, [nper, pmt, pv, fv, when])
+
     rn = guess
-    iter = 0
```
Comment (garrypolley) [inline numpy/lib/financial.py]: `iter` is a keyword, so I changed it to iterator.

Comment (eric-wieser) [reply numpy/lib/financial.py]: Not a keyword, just a builtin. Still sensible though.

File: numpy/lib/financial.py [hunk 15]
```diff
@@ -545,10 +542,10 @@ def rate(nper, pmt, pv, fv, when='end', guess=0.10, tol=1e-6, maxiter=100):
         Future value
     when : {{'begin', 1}, {'end', 0}}, {string, int}, optional
         When payments are due ('begin' (1) or 'end' (0))
-    guess : float, optional
-        Starting guess for solving the rate of interest
-    tol : float, optional
-        Required tolerance for the solution
+    guess : float/Decimal, optional
+        Starting guess for solving the rate of interest, default 0.1
+    tol : float/Decimal, optional
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: Maybe just `Number` now?

File: numpy/lib/financial.py [hunk 16]
```diff
@@ -597,6 +605,8 @@ def irr(values):
     that gives a net present value of 0.0; for a more complete explanation,
     see Notes below.
 
+    Decimal type is not supported for values.
```
Comment (eric-wieser) [inline numpy/lib/financial.py]: I wonder if `` `~decimal.Decimal` `` would link to the python docs here. @MSeifert04? - you seem to have a bunch of experience with Sphinx

**Comment by eric-wieser** [comment]
Looking pretty good now!

This will want a release note, and you should maybe add something to the docstring at the top of the file too explaining that `Decimal` is supported unless otherwise stated. There might be some other documentation to update too.

File: numpy/lib/financial.py (hunk 7 above)
Comment (eric-wieser) [reply numpy/lib/financial.py]: I might be tempted to move the `due to ...` out of the docstring and into a comment - the end user does not care why it's not supported. Then including `gh-9955` in that comment would be handy

**Comment by garrypolley** [comment]
I think I've updated the documentation as needed/desired. I'm not sure on the sphinx syntax, but from what I could google it seemed correct. 

I tried to build the docs locally to test them, but when I built them, they didn't seem to update to my new code.

File: doc/release/1.14.0-notes.rst [hunk 17]
```diff
@@ -306,6 +306,11 @@ now supported for these arrays:
  * `arr.resize(...)`
  * `pickle.dumps(arr)`
 
+Support for ``decimal.Decimal`` in ``np.financial``
```
Comment (eric-wieser) [inline doc/release/1.14.0-notes.rst]: Isn't it `np.lib.financial`?

File: numpy/lib/tests/test_financial.py [hunk 18]
```diff
@@ -83,86 +151,194 @@ def test_mirr(self):
...
-        assert_almost_equal(np.nper(0.075, -2000, 0, 100000.),
-                            np.nper(0.075, -2000, 0, 100000., 'end'), 4)
-        assert_almost_equal(np.nper(0.075, -2000, 0, 100000., 0),
-                            np.nper(0.075, -2000, 0, 100000., 'end'), 4)
+        assert_equal(np.nper(0.075, -2000, 0, 100000.),
+                     np.nper(0.075, -2000, 0, 100000., 'end'))
+        assert_equal(np.nper(0.075, -2000, 0, 100000., 0),
+                     np.nper(0.075, -2000, 0, 100000., 'end'))
+
+    def test_when_decimal(self):
```
Comment (eric-wieser) [inline numpy/lib/tests/test_financial.py]: Is this testing setting the `when` argument to a decimal? Should that be allowed?

Comment (garrypolley) [reply numpy/lib/tests/test_financial.py]: It's testing that if the `when` argument is passed, that decimals are still supported. Mildly redundant, but I wanted to make sure it was covered.

Comment (eric-wieser) [reply numpy/lib/tests/test_financial.py]: Can you make the docstring `"""Test that decimals are still supported if the when argument is passed"""`, and maybe rename to `test_decimal_with_when`, to make it clearer that `when` itself is not a decimal.

**Comment by charris** [comment]
@garrypolley We have been looking to pull the financial functions out of NumPy and put them into a separate repository. Any chance that you might be interesting in taking over such a repository?

**Review by eric-wieser — APPROVED** [review]
Looks good to me, I haven't checked the tests in detail, but it seems unlikely that the wrong results are being produced, especially if the old tests still pass. I'll leave this for a few days to see if anyone else has comments.

File: numpy/lib/tests/test_financial.py (hunk 5 above)
Comment (charris) [reply numpy/lib/tests/test_financial.py]: We will be getting rid of `unittest` in the switch to pytest.

**Comment by garrypolley** [comment]
> We have been looking to pull the financial functions out of NumPy and put them into a separate repository. Any chance that you might be interesting in taking over such a repository?

At this very  moment, it isn't likely I'd want to own them myself. I'm not 100% a no, but I also like that numpy has these functions built-in and we can pass numpy arrays/objects into it. If numpy is planning a more-module and plugin approach then I may consider it more.  

I'd mostly be afraid of it not being maintained by a strong community if I took it over just myself.

**Comment by garrypolley** [comment]
I _think_ I've resolved the merge conflict on the release file, not sure what I did wrong the first time, but the second time seems to have stuck.

**Comment by eric-wieser** [comment]
Thanks @garrypolley!
